nntrainer/app_context.h

```diff
@@ -52,9 +94,110 @@ class AppContext {
    */
   const std::string getWorkingPath(const std::string &path = "");
 
+  template <typename T>
+  const int registerFactory(const PtrFactoryType<T> factory,
+                            const std::string &key = "",
+                            const int int_key = -1) {
+    FactoryType<T> f = factory;
+    return registerFactory(f, key, int_key);
+  }
+
+  /**
+   * @brief Factory register function, use this function to register custom
+   * object
+   *
+   * @tparam T object to create
+   * @param factory factory function that creates std::unique_ptr<T>
+   * @param key key to access the factory, if key is empty, try to find key by
+   * calling factory({})->getType();
+   * @param int_key key to access the factory by integer, if it is -1(default),
+   * the function automatically unsigned the key and return
+   * @return const int unique integer value to access the current factory
+   * @throw invalid argument when key and/or int_key is already taken
+   */
+  template <typename T>
+  const int registerFactory(const FactoryType<T> factory,
+                            const std::string &key = "",
+                            const int int_key = -1) {
+
+    auto &index = std::get<IndexType<T>>(factory_map);
+    auto &str_map = std::get<StrIndexType<T>>(index);
+    auto &int_map = std::get<IntIndexType>(index);
+
+    const std::string &assigned_key = key == "" ? factory({})->getType() : key;
+
+    const std::lock_guard<std::mutex> lock(factory_mutex);
+    if (str_map.find(assigned_key) != str_map.end()) {
+      std::stringstream ss;
+      ss << "cannot register factory with already taken key: " << key;
+      throw std::invalid_argument(ss.str().c_str());
+    }
+
+    if (int_key != -1 && int_map.find(int_key) != int_map.end()) {
+      std::stringstream ss;
+      ss << "cannot register factory with already taken int key: " << int_key;
+      throw std::invalid_argument(ss.str().c_str());
+    }
+
+    int assigned_int_key = int_key == -1 ? str_map.size() + 1 : int_key;
+
+    str_map[assigned_key] = factory;
+    int_map[assigned_int_key] = assigned_key;
+
+    ml_logd("factory has registered with key: %s, int_key: %d",
+            assigned_key.c_str(), assigned_int_key);
+
+    return assigned_int_key;
+  }
+
+  template <typename T>
+  PtrType<T> createObject(const int int_key, const PropsType &props = {}) {
+    auto &index = std::get<IndexType<T>>(factory_map);
+    auto &int_map = std::get<IntIndexType>(index);
+
+    const auto &entry = int_map.find(int_key);
+
+    if (entry == int_map.end()) {
+      std::stringstream ss;
+      ss << "Int Key is not found for the object. Key: " << int_key;
+      throw exception::not_supported(ss.str().c_str());
+    }
+
+    return createObject<T>(entry->second, props);
+  }
+
+  template <typename T>
+  PtrType<T> createObject(const std::string &key, const PropsType &props = {}) {
+    auto &index = std::get<IndexType<T>>(factory_map);
```

Comment:
sorry for the basic question - but I wonder if factory_map can contain multiple tuples? as factory_map is limited to just `optimizer`.

Reply:
Yes, you can check the same part #753 that layer is added :)

Reply:
got it.